File: src/bot/commands/clan/compo.ts
import { AutocompleteInteraction, SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { EMOJIS } from '../../../util/emojis.js';
import { padStart } from '../../../util/helper.js';

/** Town Hall composition of a clan's members. */
export default class CompoCommand extends Command {
  public constructor() {
    super('compo', { category: 'clan', description: 'Show a clan’s Town Hall composition.', defer: true });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('compo')
      .setDescription('Show a clan’s Town Hall composition.')
      .addStringOption((opt) =>
        opt.setName('tag').setDescription('Clan tag or alias.').setAutocomplete(true)
      );
  }

  public async autocomplete(interaction: AutocompleteInteraction, args: Record<string, any>) {
    await interaction.respond(await this.client.resolver.clanAutocomplete(interaction.guildId, String(args.tag ?? '')));
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const { data: clan, error } = await this.client.resolver.resolveClan(interaction, args.tag);
    if (!clan) return interaction.editReply(error!);

    // Member list from the clan endpoint doesn't include TH level, so fetch each member's profile
    // in parallel (bounded by the clan size of 50).
    const players = await Promise.all(clan.members.map((m) => this.client.coc.getPlayer(m.tag)));
    const counts = new Map<number, number>();
    let total = 0;
    for (const p of players) {
      if (!p) continue;
      counts.set(p.townHallLevel, (counts.get(p.townHallLevel) ?? 0) + 1);
      total++;
    }

    const rows = [...counts.entries()]
      .sort((a, b) => b[0] - a[0])
      .map(([th, n]) => `${EMOJIS.TOWN_HALL} \`TH${padStart(th, 2)}\` — **${n}**`);

    const embed = baseEmbed(this.client, interaction)
      .setTitle(`${clan.name} (${clan.tag}) — Composition`)
      .setDescription(rows.join('\n') || '_No data._')
      .setFooter({ text: `${total} members` });

    return interaction.editReply({ embeds: [embed] });
  }
}

File: cogs/clan/commands/compo.py
"""/clan compo — Town Hall composition."""

import collections

import discord

from utils.embeds import base_embed, error_embed
from utils.emojis import E_TOWNHALL
from utils.helpers import pad_start
from utils.resolver import resolve_clan


async def compo(interaction: discord.Interaction, tag: str = None):
    await interaction.response.defer()
    try:
        clan = await resolve_clan(interaction, tag)
    except ValueError as err:
        await interaction.followup.send(embed=error_embed(str(err)))
        return

    # ClanMember carries the town hall level, so no extra per-player lookups needed.
    counts = collections.Counter(m.town_hall for m in clan.members)
    rows = [f"{E_TOWNHALL} `TH{pad_start(th, 2)}` — **{n}**" for th, n in sorted(counts.items(), reverse=True)]

    embed = await base_embed(interaction, title=f"{clan.name} ({clan.tag}) — Composition")
    embed.description = "\n".join(rows) or "_No data._"
    embed.set_footer(text=f"{clan.member_count} members")
    await interaction.followup.send(embed=embed)
